### ISSUE TITLE
[torch.compile] Massive numerical discrepancy (~0.05 Max Abs Error) in Conv+BN+SiLU fusion with INDUCTOR_FALLBACK_RANDOM=1

### ISSUE THREAD

**Issue by zifan6699** [pytorch/pytorch#175952]
### 🐛 Describe the bug

When using torch.compile with the Inductor backend and enabling INDUCTOR_FALLBACK_RANDOM=1, there is a significant numerical divergence compared to Eager mode. In a standalone reproduction involving the first three blocks of a YOLOv8-like architecture (Conv + BN + SiLU), the Max Absolute Error reaches ~0.05, and the Max Relative Error exceeds 10^5. This suggests an unstable numerical approximation within the fused kernels.
Observed Results:
Comparing Mode A (Eager Execution) vs Mode B (Compiled with INDUCTOR_FALLBACK_RANDOM=1):
Max Absolute Error: 5.4588e-02
Max Relative Error: 1.0962e+05
Rel L2 Error (Global): 3.8443e-04
Eager output range: [-0.2785, 97.1013]
Compiled output range: [-0.2785, 97.0889]
Why this matters:
While a 0.05 absolute error might seem small in isolation, this discrepancy is roughly 4 orders of magnitude higher than the expected numerical noise for FP32 fusion (typically < 1e-6). In full-scale YOLOv8 models, I have observed this error accumulating across layers, eventually reaching a Max Absolute Error of 0.4 to 0.6, which is large enough to flip classification labels and compromise model reliability.
Reproduction Script:
This script is truly standalone and does not require any external weights or data files. It manually initializes the weights to ensure intermediate activations reach a magnitude that exposes the numerical instability.

```python
import torch
import torch.nn as nn
import os
import contextlib

SEED = 0
torch.manual_seed(SEED)
torch.cuda.manual_seed_all(SEED)

@contextlib.contextmanager
def set_inductor_env_vars(opt_str):
    original_env = os.environ.copy()
    try:
        s = (opt_str or "0000")[:4].ljust(4, "0")
        os.environ["INDUCTOR_FALLBACK_RANDOM"] = "1" if s[0] == "1" else "0"
        os.environ["INDUCTOR_EPILOGUE_FUSION"] = "1" if s[1] == "1" else "0"
        os.environ["INDUCTOR_SHAPE_PADDING"] = "1" if s[2] == "1" else "0"
        os.environ["INDUCTOR_DYNAMIC"] = "1" if s[3] == "1" else "0"
        yield
    finally:
        os.environ.clear()
        os.environ.update(original_env)

class Conv(nn.Module):
    def __init__(self, in_channels, out_channels, kernel_size=3, stride=1, padding=1):
        super().__init__()
        self.conv = nn.Conv2d(in_channels, out_channels, kernel_size, stride, padding, bias=False)
        self.bn = nn.BatchNorm2d(out_channels, eps=0.001, momentum=0.03)
        self.act = nn.SiLU()

    def forward(self, x):
        return self.act(self.bn(self.conv(x)))

class Bottleneck(nn.Module):
    def __init__(self, c1, c2):
        super().__init__()
        self.cv1 = Conv(c1, c2, 3, 1, 1)
        self.cv2 = Conv(c2, c2, 3, 1, 1)

    def forward(self, x):
        return self.cv2(self.cv1(x)) + x

class C2f(nn.Module):
    def __init__(self, c1, c2, n=1, e=0.5):
        super().__init__()
        self.c = int(c2 * e)
        self.cv1 = Conv(c1, self.c, 1, 1, 0)
        self.cv2 = Conv((n + 1) * self.c, c2, 1, 1, 0)
        self.m = nn.ModuleList([Bottleneck(self.c, self.c) for _ in range(n)])

    def forward(self, x):
        y = [self.cv1(x)]
        for module in self.m:
            y.append(module(y[-1]))
        return self.cv2(torch.cat(y, 1))

class YOLOv8FirstThreeBlocks(nn.Module):
    def __init__(self):
        super().__init__()
        self.block0 = Conv(3, 16, 3, 2, 1)
        self.block1 = Conv(16, 32, 3, 2, 1)
        self.block2 = C2f(32, 32, n=1)

    def forward(self, x):
        x = self.block0(x)
        x = self.block1(x)
        x = self.block2(x)
        return x

def main():
    device = torch.device('cuda')
    
    torch.manual_seed(SEED)
    model = YOLOv8FirstThreeBlocks().to(device).eval()
    
    for name, param in model.named_parameters():
        if 'weight' in name and 'bn' not in name:
            nn.init.uniform_(param, -0.5, 0.5)
        elif 'bn' in name and 'weight' in name:
            nn.init.ones_(param)
        elif 'bn' in name and 'bias' in name:
            nn.init.zeros_(param)
    
    torch.manual_seed(SEED)
    x_in = torch.randn(64, 3, 640, 640).to(device)
    
    print("\n[Mode A] Eager Execution...")
    with torch.no_grad():
        out_eager = model(x_in)
    print("Eager mode completed")
    
    print("\n[Mode B] Compiled Execution (INDUCTOR_FALLBACK_RANDOM=1)...")
    with set_inductor_env_vars('1000'):
        compiled = torch.compile(model)
        with torch.no_grad():
            _ = compiled(x_in)
            out_compiled = compiled(x_in)
    print("Compiled mode completed")
    
    max_abs_err = torch.abs(out_eager - out_compiled).max().item()
    rel_l2_err = (torch.norm(out_eager - out_compiled) / (torch.norm(out_eager) + 1e-12)).item()
    max_rel_err = torch.abs((out_eager - out_compiled) / (torch.abs(out_eager) + 1e-12)).max().item()
    
    print("\n" + "="*100)
    print("Numerical Difference Analysis (Truly Standalone):")
    print(f"  Max Absolute Error:        {max_abs_err:.4e}")
    print(f"  Max Relative Error:        {max_rel_err:.4e}")
    print(f"  Rel L2 Error (Global):     {rel_l2_err:.4e}")
    print(f"  Eager output range:        [{out_eager.min().item():.4f}, {out_eager.max().item():.4f}]")
    print(f"  Compiled output range:     [{out_compiled.min().item():.4f}, {out_compiled.max().item():.4f}]")
    print("="*100)

if __name__ == "__main__":
    main()
```

```
[Mode A] Eager Execution...
Eager mode completed

[Mode B] Compiled Execution (INDUCTOR_FALLBACK_RANDOM=1)...
Compiled mode completed

====================================================================================================
Numerical Difference Analysis (Truly Standalone):
  Max Absolute Error:        5.4588e-02
  Max Relative Error:        1.0962e+05
  Rel L2 Error (Global):     3.8443e-04
  Eager output range:        [-0.2785, 97.1013]
  Compiled output range:     [-0.2785, 97.0889]
====================================================================================================

```

### Versions

PyTorch version: 2.10.0+cu126
Is debug build: False
CUDA used to build PyTorch: 12.6
ROCM used to build PyTorch: N/A

OS: Ubuntu 24.04.3 LTS (x86_64)
GCC version: (Ubuntu 13.3.0-6ubuntu2~24.04) 13.3.0
Clang version: Could not collect
CMake version: Could not collect
Libc version: glibc-2.39

Python version: 3.10.19 | packaged by conda-forge | (main, Jan 26 2026, 23:45:08) [GCC 14.3.0] (64-bit runtime)
Python platform: Linux-6.8.0-90-generic-x86_64-with-glibc2.39
Is CUDA available: True
CUDA runtime version: 12.6.20
CUDA_MODULE_LOADING set to:
GPU models and configuration:
GPU 0: NVIDIA GeForce RTX 3090
GPU 1: NVIDIA GeForce RTX 3090

Nvidia driver version: 560.35.03
cuDNN version: Could not collect
Is XPU available: False
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True
Caching allocator config: N/A

CPU:
Architecture: x86_64
CPU op-mode(s): 32-bit, 64-bit
Address sizes: 40 bits physical, 48 bits virtual
Byte Order: Little Endian
CPU(s): 48
On-line CPU(s) list: 0-47
Vendor ID: GenuineIntel
Model name: QEMU Virtual CPU version 2.5+
CPU family: 15
Model: 107
Thread(s) per core: 1
Core(s) per socket: 48
Socket(s): 1
Stepping: 1
BogoMIPS: 4190.15
Flags: fpu de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ht syscall nx lm constant_tsc nopl xtopology cpuid tsc_known_freq pni ssse3 fma cx16 sse4_1 sse4_2 x2apic movbe popcnt aes xsave avx f16c hypervisor lahf_lm abm cpuid_fault pti bmi1 avx2 bmi2 avx512f avx512dq avx512cd avx512bw avx512vl
Hypervisor vendor: KVM
Virtualization type: full
L1d cache: 1.5 MiB (48 instances)
L1i cache: 1.5 MiB (48 instances)
L2 cache: 192 MiB (48 instances)
L3 cache: 16 MiB (1 instance)
NUMA node(s): 1
NUMA node0 CPU(s): 0-47
Vulnerability Gather data sampling: Not affected
Vulnerability Itlb multihit: KVM: Mitigation: VMX unsupported
Vulnerability L1tf: Mitigation; PTE Inversion
Vulnerability Mds: Vulnerable: Clear CPU buffers attempted, no microcode; SMT Host state unknown
Vulnerability Meltdown: Mitigation; PTI
Vulnerability Mmio stale data: Unknown: No mitigations
Vulnerability Reg file data sampling: Not affected
Vulnerability Retbleed: Not affected
Vulnerability Spec rstack overflow: Not affected
Vulnerability Spec store bypass: Vulnerable
Vulnerability Spectre v1: Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2: Mitigation; Retpolines; STIBP disabled; RSB filling; PBRSB-eIBRS Not affected; BHI Retpoline
Vulnerability Srbds: Not affected
Vulnerability Tsx async abort: Not affected
Vulnerability Vmscape: Not affected

Versions of relevant libraries:
[pip3] numpy==1.26.4
[pip3] nvidia-cublas-cu12==12.6.4.1
[pip3] nvidia-cuda-cupti-cu12==12.6.80
[pip3] nvidia-cuda-nvrtc-cu12==12.6.77
[pip3] nvidia-cuda-runtime-cu12==12.6.77
[pip3] nvidia-cudnn-cu12==9.10.2.21
[pip3] nvidia-cufft-cu12==11.3.0.4
[pip3] nvidia-curand-cu12==10.3.7.77
[pip3] nvidia-cusolver-cu12==11.7.1.2
[pip3] nvidia-cusparse-cu12==12.5.4.2
[pip3] nvidia-cusparselt-cu12==0.7.1
[pip3] nvidia-nccl-cu12==2.27.5
[pip3] nvidia-nvjitlink-cu12==12.6.85
[pip3] nvidia-nvtx-cu12==12.6.77
[pip3] onnxruntime-gpu==1.23.2
[pip3] optree==0.18.0
[pip3] pytorch-triton==3.2.0+git4b3bb1f8
[pip3] torch==2.10.0+cu126
[pip3] torchaudio==2.11.0.dev20260127+cu126
[pip3] torchvision==0.25.0+cu126
[pip3] triton==3.6.0+git9844da95
[conda] numpy 1.26.4 pypi_0 pypi
[conda] nvidia-cublas-cu12 12.6.4.1 pypi_0 pypi
[conda] nvidia-cuda-cupti-cu12 12.6.80 pypi_0 pypi
[conda] nvidia-cuda-nvrtc-cu12 12.6.77 pypi_0 pypi
[conda] nvidia-cuda-runtime-cu12 12.6.77 pypi_0 pypi
[conda] nvidia-cudnn-cu12 9.10.2.21 pypi_0 pypi
[conda] nvidia-cufft-cu12 11.3.0.4 pypi_0 pypi
[conda] nvidia-curand-cu12 10.3.7.77 pypi_0 pypi
[conda] nvidia-cusolver-cu12 11.7.1.2 pypi_0 pypi
[conda] nvidia-cusparse-cu12 12.5.4.2 pypi_0 pypi
[conda] nvidia-cusparselt-cu12 0.7.1 pypi_0 pypi
[conda] nvidia-nccl-cu12 2.27.5 pypi_0 pypi
[conda] nvidia-nvjitlink-cu12 12.6.85 pypi_0 pypi
[conda] nvidia-nvtx-cu12 12.6.77 pypi_0 pypi
[conda] optree 0.18.0 pypi_0 pypi
[conda] pytorch-triton 3.2.0+git4b3bb1f8 pypi_0 pypi
[conda] torch 2.10.0+cu126 pypi_0 pypi
[conda] torchaudio 2.11.0.dev20260127+cu126 pypi_0 pypi
[conda] torchvision 0.25.0+cu126 pypi_0 pypi
[conda] triton 3.6.0+git9844da95 pypi_0 pypi

cc @ezyang @gchanan @kadeng @msaroufim @chauhang @penguinwu @voznesenskym @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @muchulee8 @amjames @aakhundov @coconutruben @jataylo
